Adding Speaker Status and Speaker Title in Events Form

akash19coder opened the pull request:
<!-- If your PR fixes an open issue, use `Closes #999` to link your PR with the issue. #999 stands for the issue number you are fixing -->

## Fixes Issue
Closes #7820 

<!-- Remove this section if not applicable -->

<!-- Example: Closes #31 -->

## Changes proposed
 **1. Added two inputs field in the events form namely: Are you speaking? and What topic are you speaking on?**

![Screenshot (23)](https://github.com/EddieHubCommunity/LinkFree/assets/72060440/7d746af8-5a8f-42ba-9935-b2db04c65f14)

** 2. User profile to display speaking information** 
![Screenshot (22)](https://github.com/EddieHubCommunity/LinkFree/assets/72060440/57270824-4d7c-4a93-a6c9-895f1f14db7e)

<!-- List all the proposed changes in your PR -->

<!-- Mark all the applicable boxes. To mark the box as done follow the following conventions -->
<!--
[x] - Correct; marked as done
[X] - Correct; marked as done

[ ] - Not correct; marked as **not** done
-->

## Check List (Check all the applicable boxes) <!-- Follow the above conventions to check the box -->

- [x] My code follows the code style of this project.
- [ ] My change requires changes to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] All new and existing tests passed.
- [x] This PR does not contain plagiarized content.
- [x] The title of my pull request is a short description of the requested changes.

## Screenshots

<!-- Add all the screenshots which support your changes -->

## Note to reviewers

<!-- Add notes to reviewers if applicable -->


<a href="https://gitpod.io/#https://github.com/EddieHubCommunity/LinkFree/pull/8006"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

eddiejaoude reviewed (changes requested):
Thank you for the contribution 👍 

A few things I noticed:
- too many strings being used, we should move these to config
- json events have this info and needs to be loaded into existing events

akash19coder commented:
> * json events have this info and needs to be loaded into existing events

eddie what json events are we referring to could you please give me some more context. I actually didn't understand

eddiejaoude commented:
> > * json events have this info and needs to be loaded into existing events
> 
> eddie what json events are we referring to could you please give me some more context. I actually didn't understand

The reload script will need updating for existing events managed by the json files https://github.com/EddieHubCommunity/LinkFree/blob/main/pages/api/system/reload.js

eddiejaoude commented on models/Profile/Event.js:
```diff
@@ -34,6 +34,11 @@ const EventSchema = new Schema({
   },
+  userStatus: {
+    type: String,
+    enum: ["speaking", "not-speaking"],
```
I think the options should be `speaking`, `attending`

eddiejaoude commented on pages/account/manage/event/[[...data]].js:
```diff
@@ -69,6 +70,14 @@ export default function ManageEvent({ BASE_URL, event }) {
   const [endDate, setEndDate] = useState(
     event.date?.end && formatDate(event.date?.end)
   );
+  const DropDownOptions = ["Non-speaking","Speaking"];
```
Duplicate again, please set in config and reuse

akash19coder replied:
Removed duplicacy:
```
  const [endDate, setEndDate] = useState(
    event.date?.end && formatDate(event.date?.end)
  );
  const DropDownOptions = config.events.userStatus;
```

eddiejaoude commented on pages/account/manage/event/[[...data]].js:
```diff
@@ -69,6 +70,14 @@ export default function ManageEvent({ BASE_URL, event }) {
   );
+  const DropDownOptions = ["Non-speaking","Speaking"];
+  const [userStatus, setuserStatus] = useState(event.userStatus || "");
+  const [speakingTopic, setspeakingTopic] = useState(event.speakingTopic || "Your speaking topic");
```
Default should be an empty string, use html placeholder for suggestions

akash19coder replied:
Added default to be an empty string:

`const [speakingTopic, setspeakingTopic] = useState(event.speakingTopic || "");`

Added placeholder in the component:
```
 <Input
   type="text"
   name="topic"
   label="What topic are you speaking on ?"
   placeholder= "Your speaking topic"
   onChange={(e) => setspeakingTopic(e.target.value)}
   value={speakingTopic}
   required
   maxLength="256"
  />
```

eddiejaoude commented on pages/account/manage/event/[[...data]].js:
```diff
@@ -238,6 +249,35 @@ export default function ManageEvent({ BASE_URL, event }) {
                   </div>
+                  <div className="mt-1 sm:col-span-2 sm:mt-0">
+                  <div className="mt-1">
+                        <Select
```
Formatting doesn't look correct here, Prettier VScode plugin will fix

eddiejaoude commented on models/Profile/Event.js:
```diff
@@ -34,6 +34,11 @@ const EventSchema = new Schema({
+    type: String,
+    enum: ["speaking", "not-speaking"],
+  },
+  speakingTopic: String,
```
This needs a max, maybe 256?

akash19coder replied:
Here:
```
speakingTopic: {
    type: String,
    max: 256
  }
```

eddiejaoude replied:
Please make the change in your PR, it is easier to review

eddiejaoude replied:
You marked this as resolved but the changes haven't been made, any reason why?

akash19coder replied:
I have added the field in a new commit

eddiejaoude commented on pages/account/manage/event/[[...data]].js:
```diff
@@ -238,6 +249,35 @@ export default function ManageEvent({ BASE_URL, event }) {
+                      <div className="mt-1 sm:col-span-2 sm:mt-0">
+                        <Input
+                          type="text"
+                          min="0"
```
min 0 is not recommended, if required, used `required`

also a `max` is required, this should match the model max

akash19coder replied:
Here is the change:
```
                        <Input
                          type="text"
                          name="topic"
                          label="What topic are you speaking on ?"
                          placeholder= "Your speaking topic"
                          onChange={(e) => setspeakingTopic(e.target.value)}
                          value={speakingTopic}
                          required
                          maxLength="256"
                        />
```

eddiejaoude replied:
Is it required though? What if they are not speaking?

akash19coder replied:
If the user selects the speaking option it only appears then else doesn't appear

eddiejaoude commented on pages/account/manage/event/[[...data]].js:
```diff
@@ -144,7 +155,7 @@ export default function ManageEvent({ BASE_URL, event }) {
     <>
       <PageHead
-        title="Manage Milstone"
+        title="Manage Milestone"
```
Good spot 👍

eddiejaoude commented on config/app.json:
```diff
@@ -6,6 +6,7 @@
     "inline",
     "column-2"
   ],
+  "userStatus":"Speaking",
```
This is not clear, it just seems randomly added, I think it needs to go into an `event` object

akash19coder replied:
This is how I have represented, is it fine? 
```
"layouts": [
    "classic",
    "inline",
    "column-2"
  ],
  "events":{
    "userStatus":["Attending","Speaking"]
  },
  "isr": {
    "homepage": {
      "revalidateSeconds": 900,
      "description": "15 * 60 = 15 minutes"
    },
```

eddiejaoude replied:
Yes this is better 👍 thank you

akash19coder commented on components/event/EventCard.js:
```diff
@@ -75,11 +76,11 @@ export default function EventCard({ manage, event, usernames }) {
-                    {event.userStatus == "speaking" && " at "} this event
-                    {event.userStatus == "speaking" && event?.speakingTopic && (
+                    {event.userStatus == config.userStatus && " at "} this event
+                    {event.userStatus == config.userStatus && event?.speakingTopic && (
```
Based on new config updation i.e : 
 ```
 "layouts": [
    "classic",
    "inline",
    "column-2"
  ],
  "events":{
    "userStatus":["Attending","Speaking"]
  },
```
Here is the change, what do you think Eddie?
```
             {event.userStatus == config.events.userStatus[1] && (
                  <div className="text-primary-medium-low dark:text-primary-low-medium italic hidden lg:block">
                    {event.userStatus}
                    {event.userStatus == config.events.userStatus[1] && " at "} this event
                    {event.userStatus == config.events.userStatus[1] && event?.speakingTopic && (
                      <>
                        {" "}
                        on <b>{event.speakingTopic}</b>
                      </>
                    )}
                  </div>
```

akash19coder commented:
If all the changes are alright will commit the changes and push. 😄

eddiejaoude commented:
Thank you for the replies, please can you commit the changes so it is easier to review 

Also, can you resolve conflicts, thank you 👍

eddiejaoude commented on config/app.json:
```diff
@@ -6,6 +6,9 @@
     "column-2"
   ],
+  "events":{
+    "userStatus":["Attending","Speaking"]
```
Looking at the usage code from here, it is still not readable or maintainable

```js
event.userStatus == config.events.userStatus[1]
```

So I suggest using an object not an array

```suggestion
    "userStatus": { attending: "Attending", speaking: "Speaking"}
```

Then the usage will become

```js
event.userStatus == config.events.speaking
```

eddiejaoude commented:
@akash19coder you keep merging main into your branch, however you haven't addressed some comments still but you keep chasing me?

akash19coder commented:
I was so desperate for the PR to merge that I overlooked the mistake. I am sorry.

> @akash19coder you keep merging main into your branch, however you haven't addressed some comments still but you keep chasing me?

I was so desperate for the PR to merge that I overlooked the mistake. my bad. 🙂

akash19coder commented:
@eddiejaoude I haven't made any changes in UserEvent.js or event.js file. However, when I create events it says the event was created successfully but doesn't load the event in the event tab. It was working fine before it stopped working. I wonder if you mind having a look at it

dan-mba commented on models/Profile/Event.js:
```diff
@@ -34,6 +36,14 @@ const EventSchema = new Schema({
     startingFrom: Number,
   },
+  userStatus: {
+    type: String,
```
Why is this a string and not a boolean?
There are only 2 states:
attending & not speaking
attending & speaking

This is the perfect use case for a boolean

eddiejaoude replied:
Could there be more? What about making it an object with ...?

- `isSpeaking`
- `isAttending`

dan-mba replied:
I was under the assumption that people were only posting events they are attending. That would make isAttending always true. Is this not the case?

eddiejaoude replied:
Yeah I think people are posting events they are aware of or want to support - which I think is a good thing, I am keen to have a central place for events

dan-mba reviewed (commented):
I believe switching to a boolean will make for cleaner code.
The UI can then be improved by using the Toggle component for speaking.
@eddiejaoude Your thoughts?

dan-mba commented on pages/account/manage/event/[[...data]].js:
```diff
@@ -102,19 +111,21 @@ export default function ManageEvent({ BASE_URL, event }) {
     e.preventDefault();
 
-    let alert = "created";
+let alert = "created";
```
please fix the indenting change

dan-mba commented on pages/account/manage/event/[[...data]].js:
```diff
@@ -263,10 +275,34 @@ export default function ManageEvent({ BASE_URL, event }) {
                     <Input
                       type="number"
-                      min="0"
+                      minLength="0"
```
This should be `min` to prevent negative prices

dan-mba commented on components/event/EventCard.js:
```diff
@@ -66,11 +66,11 @@ export default function EventCard({ manage, event, usernames }) {
-                    {event.userStatus == "speaking" && event?.speakingTopic && (
+                    {event.isSpeaking}
+                    {"Speaking at "} this event
+                    {event.isSpeaking  && event?.speakingTopic && (
```
The `event.isSpeaking` check is unnecessary as this section is only displayed when this is true.

akash19coder commented on pages/api/system/reload.js:
```diff
@@ -261,6 +261,8 @@ export default async function handler(req, res) {
                   end: event.date.end,
                 },
                 url: event.url,
+                isSpeaking: event.isSpeaking,
```
@dan-mba  we don't have userStatus field... I am a bit confused why we are adding that?

dan-mba reviewed (approved):
Looks good to me

eddiejaoude reviewed (approved):
Great collaboration, thank you all 👍
